Add Dockerfile

arjxn-py opened the pull request:
Still a work in progress.
This PR aims to add a Dockerfile to containerize the package & publish Docker images of the same.

- [x] Can be GLIBC-based
- [x] Needs to derive from official Python images
- [x] Should be multi-architecture
- [x] Should come with a C compiler (GCC/Clang)
- [x] Should come with Git installed
- [x] Should come with Go installed (same version as upstream)
- [x] Should come with Zig installed
- [x] Building & testing of Image in CI

Related to #74

arjxn-py commented:
Brief Update - A preliminary Dockerfile has been set up, Parameterization of Dockerfile for dynamic version installation remains, plus there are some issues while activating venv, I'm on it :)

agriyakhetarpal commented:
Sure, could we skip the venv? I think we do not plan to add any external dependencies from PyPI, so we won't ever need the utilities of the isolation that a virtual environment provides, because the Docker container would provide that. Installing straight with the system `pip` should be alright.

arjxn-py commented:
> I'm not sure I follow. Do you mean installing multiple Hugo versions? 

No no, by this I meant about the version of `go` & `zig`. I should have specified that above 😅

agriyakhetarpal commented:
Fair enough. I don't have any release automation for bumping versions set up right now given how minimal the entire project is, so it will be okay to keep the Zig and Go tarballs hardcoded (as they currently are).

agriyakhetarpal commented:
We can't really control the C compiler that we receive from the system package manager. However, I determine the Go toolchain version from the notes for every Hugo release. That is more important to keep in sync because of security concerns; sometimes some CVEs are fixed – so I make sure to do so in all relevant files when bumping the version.

agriyakhetarpal commented:
Through #81 I added some non-GitHub Linux aarch64/arm64 runners (the capacity is limited to 800 minutes per month or so, as I am on 5$ of free credit as a trial). Feel free to use them in a workflow or two here in your PR (won't work on your fork); I would happily approve workflow runs for them. It will be much better than running QEMU and waiting for 40 minutes to build.

arjxn-py commented:
> Through #81 I added some non-GitHub Linux aarch64/arm64 runners (the capacity is limited to 800 minutes per month or so, as I am on 5$ of free credit as a trial). Feel free to use them in a workflow or two here in your PR (won't work on your fork); I would happily approve workflow runs for them. It will be much better than running QEMU and waiting for 40 minutes to build.

Thanks for this @agriyakhetarpal, I am in the process to define +1 workflow for Docker images hence it is nicer to use aarch64/arm64 runners instead, I anticipate that the capacity wouldn't be an issue at the moment as it takes ~10 minutes to build the image for me locally.
Plus, I've gotten rid of venv for now as you suggested and added tarballs hardcoded.

agriyakhetarpal commented on Dockerfile:
```diff
@@ -0,0 +1,38 @@
+FROM python:3.11.9-slim
+
+ENV GO_VERSION=1.22.1
```
```suggestion
ENV GO_VERSION=1.22.2
```

arjxn-py replied:
Bumped along in [`494673d`](https://github.com/agriyakhetarpal/hugo-python-distributions/pull/75/commits/494673ddf955125707f02a5078939e74ac3c7f78)

agriyakhetarpal reviewed (changes requested):
Thanks for writing this out, @arjxn-py, much appreciated! I had initially planned to publish these Docker images when I opened the issue, yes – however, I now am backing away from it for two reasons and some of my opinions about this have changed:

1. Hugo has Docker images already where it is being distributed, and those run without the need for a Python base image (considering that Hugo is a standalone binary and does not necessarily need a Python wrapper around it). Those will be much more reliable, and I would myself suggest people use those instead of ours, and use our built wheels instead if needed (also because the current Dockerfile targets only a stable release and not a development build)
2. Where I think these images would be more useful would be for something like getting access to a development build on a Linux runtime while being on a Windows one (where setting up compilers and other things is a bit tricky) or on any other platforms where Docker images can work. Or, for building MUSL-linked static binaries for Hugo extended via Zig using an Alpine base image (which is something that even the upstream Hugo project does not do right now, they link against GLIBC for those binaries, and the non-extended version is statically linked I think).

Therefore, I would suggest building a MUSL image for both architectures with separate Dockerfiles (the current implementation is amd64 specific so it would not have worked in the `docker/build-push-action@v5` step) and using Zig to compile it with static linkage (some instructions for this are available in the documentation which you can use). This way, we can remove the Golang installation in the Docker images too and keep everything quite minimal. Publishing the images could be looked at some later time when we would have tested these images properly, but personally I think we shouldn't release them at all since the original plan of the project was to provide isolation and easy management of Hugo binaries, which remains largely accomplished now with tools like `pipx` and Python virtual environments.

agriyakhetarpal commented:
Further context: I had previously tried MUSL wheels in PRs related to #91 and received a few segmentation faults. I didn't do a debug build and look at them in detail, though. Maybe we'll need to change `BUILD_FOR_WINDOWS` to `BUILD_STATIC` or something to be able to build a static library. If that doesn't work, we can publish a GLIBC-based image.

agriyakhetarpal commented on setup.py:
```diff
@@ -196,11 +196,11 @@ def run(self):
         # i.e., libgcc_s_seh-1.dll, libstdc++-6.dll, etc.
-        BUILDING_FOR_WINDOWS = (
+        BUILD_STATIC = (
             os.environ.get("GOOS") == "windows" or sys.platform == "win32"
```
This condition can now be modified to check if we are building for MUSL. There is just one way to do that, which is to check if the `CC` and `CXX` environment variable strings contain `linux-musl`, e.g., if CC is set to "zig cc -target x86_64-linux-musl", and so on.

arjxn-py replied:
@agriyakhetarpal I've updated the condition, does it need to check if CC is set to "zig cc -target x86_64-linux-musl" exactly?

agriyakhetarpal replied:
It could be `aarch64-linux-musl` as well, and not just `x86_64`, so we shouldn't need to check for an exact match. This commit should be okay.

agriyakhetarpal reviewed (changes requested):
Thanks, @arjxn-py! Let me test the Dockerfile locally to see if the MUSL builds are working, but for now, maybe we need to think about the predicament imposed by the `docker.yml` workflow. I have thought that the Dockerfile should be a method to build any released versions of Hugo from source or for those running MUSL systems that need a working build, but I'll still need to think about whether this should be published to Docker Hub or not. Hugo has had a recent PR for official images that could be accepted by the maintainers: https://github.com/gohugoio/hugo/pull/12573, and in general everything would be better shaped if an official image is decided upon (because there have been other images in the past that gained much more popularity than the official images, see https://github.com/gohugoio/hugo/issues/10760 but maintenance is an issue).

We should still modify the workflow here to test the `hugo version` and `hugo env` commands inside the built Docker image. We shouldn't push it (so maybe change to `push: false`).

If the MUSL builds don't work, then it's probably best to switch to a glibc-based Python image as discussed above.

arjxn-py commented on setup.py:
```diff
@@ -196,11 +196,11 @@ def run(self):
-        )
+        BUILD_STATIC = "linux-musl" in os.environ.get(
+            "CC", ""
+        ) or "linux-musl" in os.environ.get("CXX", "")
```
Done in [`d703210`](https://github.com/agriyakhetarpal/hugo-python-distributions/pull/75/commits/d703210c9c1f4ff30463b9a2040a9574d2234fe6)

agriyakhetarpal commented:
I think I will rely on https://github.com/gohugoio/hugo/issues/10760, which has now been closed through https://github.com/gohugoio/hugo/pull/12734 for the official releases of Hugo, so I'm not sure if we need this implementation or need to publish our own images. We can provide these for development purposes, of course, so I'll keep this PR open if anyone ever needs this functionality – it's only useful for cross-compilation or testing Linux stuff locally on a non-Linux machine, and compilers are readily available on all platforms (whether through conda-forge/miniconda, MinGW, Homebrew, etc.) – I think that sort of stuff could be nice to try in CI

agriyakhetarpal commented:
As we discussed earlier, I'll be closing this as not needed, but thank you for the effort, @arjxn-py!
